### ISSUE TITLE
Pre-1.21 release: ContourConfiguration and ContourDeployment CRD review

### ISSUE THREAD

**Issue by sunjayBhatia** [projectcontour/contour#4062]
This is a holding issue to document that we should do a pre-1.21 release review to ensure we are happy with the state of the new ContourConfiguration and ContourDeployment CRDs.

Steve S did some great work to get these CRDs going and we've had some in-person discussions on how we might future proof things from as they are initially implemented, restructure, follow best practices, etc. Since some of the implementation has started as well, we are getting a better sense of how the design will work in practice. Even though these CRDs are in the v1alpha1 group and we have had a design review process, we should ensure we get to as close to a final state as we can at first so we do not have to change things drastically down the line.

Things we discussed we should review include:
- kubebuilder validations (there were discussions on how to do some of these things and various people had ideas to try to propose etc.)
- in-code validation
- boolean field naming (defaulting to secure, polarity of field names e.g. "DisableFoo" vs. "EnableFoo" consistency)
- other defaulting (defaulting to secure, recommended options, etc.)
- general CRD structure, grouping of fields etc.
- etc.

My suggestion is each maintainer does a review of the API and we hold the release until we all comment with an approval

**Comment by youngnick** [comment]
Great idea, thanks @sunjayBhatia. Totally agreed, this is very similar to how upstream Kubernetes does API reviews.

**Comment by youngnick** [comment]
I haven't responded to this before now, but I think these are all great points and I think that we should address them. I'm not sure what to do yet though(which is why I hadn't repsonded until now). Thanks for the review @tsaarni.

**Comment by sunjayBhatia** [comment]
> 2. When setting single field manually e.g. spec.envoy.metrics I got error that field spec.envoy.cluster must be set. The default were not applied anymore. I believe the reason is that the cluster defaults are defined only within field spec.envoy up in the hierarchy and not repeated again on spec.envoy.cluster. Action: All defaults need to be repeated consistently on each level of the struct hierarchy.

Having an issue with this as well creating a resource from Go rather than just YAML

b/c some fields are 'optional' but not pointers+tagged with `omitempty`, they are filled in with empty go structs rather than a `nil` so they are always supplied with zero values

e.g. a configuration struct like:

```
		contourConfig = &contour_api_v1alpha1.ContourConfiguration{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "incluster-test",
				Namespace: f.Deployment.Namespace.Name,
			},
		}
```


results in the error:

```
ContourConfiguration.projectcontour.io "incluster-test" is invalid: [spec.health.address: Invalid value: "": spec.health.address in body should be at least 1 chars long, spec.debug.logLevel: Unsupported value: "":
supported values: "info", "debug", spec.metrics.address: Invalid value: "": spec.metrics.address in body should be at least 1 chars long, spec.xdsServer.type: Unsupported value: "": supported values: "contour", "envoy", spec.xdsServer.add
ress: Invalid value: "": spec.xdsServer.address in body should be at least 1 chars long, spec.envoy.health.address: Invalid value: "": spec.envoy.health.address in body should be at least 1 chars long, spec.envoy.http.address: Invalid val
ue: "": spec.envoy.http.address in body should be at least 1 chars long, spec.envoy.logging.accessLogFormat: Unsupported value: "": supported values: "envoy", "json", spec.envoy.listener.tls.minimumProtocolVersion: Unsupported value: "":
supported values: "1.2", "1.3", spec.envoy.listener.tls.cipherSuites: Required value, spec.envoy.cluster.dnsLookupFamily: Unsupported value: "": supported values: "auto", "v4", "v6", spec.envoy.metrics.address: Invalid value: "": spec.env
oy.metrics.address in body should be at least 1 chars long, spec.envoy.https.address: Invalid value: "": spec.envoy.https.address in body should be at least 1 chars long, spec.envoy.defaultHTTPVersions: Required value]
```

I think we need to make all the optional fields pointers, otherwise the defaulting doesnt seem to work when using Go

we can also possibly revisit the field defaults vs struct defaults to see what works

**Comment by tsaarni** [comment]
For all those reasons given, I've started to think that defaulting is deeply problematic feature :( I would also "vote" for pointers.

**Comment by youngnick** [comment]
I had missed this, but yes, all optional fields that are structs must have omitempty and be pointers for everything to work.

**Comment by skriss** [comment]
Is it time to just go back to embedding the defaults within Contour code, vs. trying to use API defaults? Seems like we avoid many pitfalls that way. We could still document the internal Contour defaults in the API's godoc, for visibility to users.

**Comment by youngnick** [comment]
I think that we should do defaults in the CRDs where we can - the "pointers and omitempty for optional struct fields" is a *very* unclear part of the Kubernetes API conventions and makes a lot of this really suck. As I said on another thing, there are ways to get around having to specify *all* the defaults at every level, if we are careful, but they are fiddly.

We're going to have to deal with this for Gateway API anyway, I feel like doing it for HTTPProxy helps us use that muscle.

This isn't a strong preference though, I'm definitely interested in talking more about this.

**Comment by tsaarni** [comment]
Setting all defaults at all levels solves only problem (2)  in https://github.com/projectcontour/contour/issues/4062#issuecomment-951768078 but what is your opinion about the remaining problems? 

For me these ones really seem to impact maintainability:
* (3) Defaults can't be changed reliably - Defaults are applied at the time of CR instantiation, they are coming from the CRD version that was used at that point in time.  For example, default ciphers and default timeouts are something that have been changing during time, but as far as I can see, it cannot be done anymore due to CRD defaulting. I think users expect defaults to come from the software release and having "defaults" stored persistently in their existing CR instances is unexpected.
* (5) Defaulted fields cannot be reliably added in new versions of CRD - For backwards compatibility, the general rule has been that all newly introduced attributes should be optional and have defaults. But how are the defaults applied to existing CRs at the time of CRD upgrade?

**Comment by youngnick** [comment]
I think that we need to be a little considered in how we do defaults - I think that your point about ciphers is a good one, it makes sense to have that not defaulted in the CRD so that we can change the defaults.

Also, when adding new fields to the spec, they will be set to empty on older objects, yes, so the defaults won't apply to them, good point.

I think that defaulting is most useful for things that we want to see reflected in the configured object, without needing to go look up the docs. I'm thinking default log levels, leader election details, and so on, but I think my original idea of having *as much as possible* defaulted isn't viable.

**Comment by youngnick** [comment]
I spent some more time thinking and playing with this today - this is very interesting for anyone who uses CRD defaulting, and there isn't much good tribal knowledge about it yet.

Originally, the plan was for this to allow you to use the following object:

```yaml
apiVersion: projectcontour.io/v1alpha1
kind: ContourConfiguration
metadata:
  namespace: projectcontour
  name: blank
spec: {}
```
And get the config that Contour will default to. This makes the config CRD the canonical record of Contour's settings. So you only need to set the settings you want to change, and the config object will show you what other defaults you got.

I agree that what happens on upgrades is unclear; I'm going to do some testing on that tomorrow.

What I did test today is how you can be more clear about the aim (this bit is long, sorry, but I've taken out the bits except for the relevant ones of the examples):

First, I applied the `main` CRDs

Then, I applied this:
```yaml
apiVersion: projectcontour.io/v1alpha1
kind: ContourConfiguration
metadata:
  namespace: projectcontour
  name: blank
spec: {}
```

I got this:
```yaml
apiVersion: projectcontour.io/v1alpha1
kind: ContourConfiguration
metadata:
  annotations:
  creationTimestamp: "2022-01-18T05:51:07Z"
  generation: 1
  name: ratelimit
  namespace: projectcontour
  resourceVersion: "456591250"
  uid: 842e0d6c-3a05-4e6b-bbc1-f4f09d6451a8
spec:
  # irrelevant parts removed
  xdsServer:
    address: 0.0.0.0
    port: 8001
    tls:
      caFile: /certs/ca.crt
      certFile: /certs/tls.crt
      insecure: false
      keyFile: /certs/tls.key
    type: contour
```

This uses the "specify the full tree of objects at the top level" approach.

Then, I changed the default for XDSServerConfig to only be:

```go
	// XDSServer contains parameters for the xDS server.
	// +optional
	// +kubebuilder:default={type: "contour"}
	XDSServer XDSServerConfig `json:"xdsServer"`
```

Then, updated the XDSServerConfig to add defaults (there were no defaults previously):

```go
type XDSServerConfig struct {
	// Defines the XDSServer to use for `contour serve`.
	// +kubebuilder:validation:Enum=contour;envoy
	// +kubebuilder:default=contour
	// +optional
	Type XDSServerType `json:"type"`

	// Defines the xDS gRPC API address which Contour will serve.
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:default="0.0.0.0"
	// +optional
	Address string `json:"address"`

	// Defines the xDS gRPC API port which Contour will serve.
	// +kubebuilder:default=9001
	// +optional
	Port int `json:"port"`

	// TLS holds TLS file config details.
	// +optional
	TLS *TLS `json:"tls,omitempty"`
}
```
(We will want to do this again for the TLS struct as well).

This means that the object will be created (because the `type` is set to `contour`, but the defaults will be set to whatever they are in the defaults for the inner struct).

I applied the same blank config and got:
```yaml
apiVersion: projectcontour.io/v1alpha1
kind: ContourConfiguration
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"projectcontour.io/v1alpha1","kind":"ContourConfiguration","metadata":{"annotations":{},"name":"blank","namespace":"projectcontour"},"spec":{}}
  creationTimestamp: "2022-01-18T05:59:42Z"
  generation: 1
  name: blank
  namespace: projectcontour
  resourceVersion: "456595963"
  uid: f4940ae7-4206-4134-95e8-3467e821e279
spec:
  # Extra bits removed again
  xdsServer:
    address: 0.0.0.0
    port: 9001
    type: contour
```
So, same outcome, but much easier to maintain.

This will also definitely fix problem 2 in @tsaarni's original comments.

I'm going to do some checking about what happens if you update the spec and have objects stored. I'm pretty confident that already-set fields won't be changed, but I'm not sure what will happen with fields that didn't exist before. I *think* that @tsaarni is correct, and those will be empty/unset, and not have values, but I would like to check.

If what I've said above is the case, then that leads me to think that we should use defaults sparingly, for things that we're confident we have correct (or close to). `xdsServer` is probably a good candidate, as would be `debug`, and similar fields that we're porting over from the current config file.

None of this helps to solve the defaults-not-present-in-docs problem, I'll check what we do for Gateway API.

**Comment by tsaarni** [comment]
> Also, when adding new fields to the spec, they will be set to empty on older objects, yes, so the defaults won't apply to them, good point.

and since on the Go side "empty" values will appear to be set to 0 or "", those would need to be treated as signal values that signifying empty. As it is of course well known, the generic way to separate values that are unset vs intentionally set to 0 or "" would be the use of pointers.

> If what I've said above is the case, then that leads me to think that we should use defaults sparingly, for things that we're confident we have correct (or close to). xdsServer is probably a good candidate

I wonder if it will be tricky to decide what defaults can be safely "persisted" to users configs? For example, lets say we defaulted `xdsServer` parameters like above, but later we want to deprecate and remove `contour` server type support completely. Users who likely did not even know about server types would manually need to go and expilicitely switch to `envoy` when upgrading to a release that removed support for `contour` type.

**Comment by youngnick** [comment]
> I'm going to do some checking about what happens if you update the spec and have objects stored. I'm pretty confident that already-set fields won't be changed (this is the TLS Ciphersuite problem), but I'm not sure what will happen with fields that didn't exist before (this is the upgrade problem). I _think_ that @tsaarni is correct, and those will be empty/unset, and not have values, but I would like to check.

So, I just did this:
- Have `blank` object, originally asked for as above, *without* defaults set for TLS. The object looked as above:
  ```yaml
  spec:
    # Extra bits removed again
    xdsServer:
      address: 0.0.0.0
      port: 9001
      type: contour
  ```
- Add in defaults for the TLS struct:
  ```go
  type TLS struct {
	// CA filename.
	// +optional
	// +kubebuilder:default="/certs/ca.crt"
	CAFile string `json:"caFile,omitempty"`

	// Client certificate filename.
	// +optional
	// +kubebuilder:default="/certs/tls.crt"
	CertFile string `json:"certFile,omitempty"`

	// Client key filename.
	// +optional
	// +kubebuilder:default="/certs/tls.key"
	KeyFile string `json:"keyFile,omitempty"`

	// Allow serving the xDS gRPC API without TLS.
	Insecure bool `json:"insecure"`
  }
  ```
  and at the `xdsServer` level:
  ```go
  type XDSServerConfig struct {
	// snip other details

	// TLS holds TLS file config details.
	// +optional
	// +kubebuilder:default={insecure: false}
	TLS *TLS `json:"tls,omitempty"`
  }
  ```
Then applied a `blank2` resource, same as the `blank` one, and got back:
```yaml
apiVersion: projectcontour.io/v1alpha1
kind: ContourConfiguration
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"projectcontour.io/v1alpha1","kind":"ContourConfiguration","metadata":{"annotations":{},"name":"blank2","namespace":"projectcontour"},"spec":{}}
  creationTimestamp: "2022-01-19T04:06:40Z"
  generation: 1
  name: blank2
  namespace: projectcontour
  resourceVersion: "457322683"
  uid: f73dd5c8-6370-46d5-b207-f6b8790bb3f7
spec:
  # goodbye extras
  xdsServer:
    address: 0.0.0.0
    port: 9001
    tls:
      caFile: /certs/ca.crt
      certFile: /certs/tls.crt
      insecure: false
      keyFile: /certs/tls.key
    type: contour
```

*But*, I also checked the `blank` resource, and:
```yaml
apiVersion: projectcontour.io/v1alpha1
kind: ContourConfiguration
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"projectcontour.io/v1alpha1","kind":"ContourConfiguration","metadata":{"annotations":{},"name":"blank","namespace":"projectcontour"},"spec":{}}
  creationTimestamp: "2022-01-18T05:59:42Z"
  generation: 1
  name: blank
  namespace: projectcontour
  resourceVersion: "456595963"
  uid: f4940ae7-4206-4134-95e8-3467e821e279
spec:
  # snip
  xdsServer:
    address: 0.0.0.0
    port: 9001
    tls:
      caFile: /certs/ca.crt
      certFile: /certs/tls.crt
      insecure: false
      keyFile: /certs/tls.key
    type: contour
```

This means that, if we update the APIserver defauls, and the object didn't have any stored, it will have them added. This handles the "what happens on upgrade" question.

**Comment by youngnick** [comment]
Then, I updated the definitions for the TLS struct:
```go
	// CA filename.
	// +optional
	// +kubebuilder:default="/certs/ca.crt1"
	CAFile string `json:"caFile,omitempty"`

	// Client certificate filename.
	// +optional
	// +kubebuilder:default="/certs/tls.crt1"
	CertFile string `json:"certFile,omitempty"`

	// Client key filename.
	// +optional
	// +kubebuilder:default="/certs/tls.key1"
	KeyFile string `json:"keyFile,omitempty"`
```
(just added a `1` to the end)

And I got:
```yaml
apiVersion: projectcontour.io/v1alpha1
kind: ContourConfiguration
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"projectcontour.io/v1alpha1","kind":"ContourConfiguration","metadata":{"annotations":{},"name":"blank","namespace":"projectcontour"},"spec":{}}
  creationTimestamp: "2022-01-18T05:59:42Z"
  generation: 1
  name: blank
  namespace: projectcontour
  resourceVersion: "456595963"
  uid: f4940ae7-4206-4134-95e8-3467e821e279
spec:
  # snip
  xdsServer:
    address: 0.0.0.0
    port: 9001
    tls:
      caFile: /certs/ca.crt1
      certFile: /certs/tls.crt1
      insecure: false
      keyFile: /certs/tls.key1
    type: contour
```

What this implies to me is that the actual object that's being stored is the blank one, and the defaults are being filled in for display purposes when you output it.

This means that we would get updates, *as long as they don't pull-and-reapply* the object.

I verified this as well. Once the object has a value in the actual storage, that's what you have.

It's neat but probably not worth relying on - we would still need to tell people to update the values.

**Comment by youngnick** [comment]
> I wonder if it will be tricky to decide what defaults can be safely "persisted" to users configs? For example, lets say we defaulted `xdsServer` parameters like above, but later we want to deprecate and remove `contour` server type support completely. Users who likely did not even know about server types would manually need to set `envoy` before upgrading.

Changing the values for an `enum` field is actually an API-breaking change and requires an API revision bump by the API conventions. (this is why you don't see a lot of `enum` fields in the core Kubernetes API). What we did here for the Gateway API is make a typed string field and supply constants for the "enum" values. That way, your controller can say "I support these values, the value `whatever` is not supported".

In our case here, I think we should use a typed string with supplied constants. Then, our deprecation cycle goes like this:
- (warning) Your `xdsServer.type` value of `contour` is deprecated and will be removed. Please change it to prevent future issues. 
- (fatal) `xdsServer.type` value of `contour` is not supported. Currently supported values: `envoy`. 

I'm writing up notes on all of these API fiddly bits so I can put them somewhere for future reference, because I don't think a lot of them are written down in the API conventions, even.

**Comment by youngnick** [comment]
Okay, I had more of a look over this, and I think these are my overall recommendations:

**We should either have everything be optional, with apiserver defaults for things that we don't anticipate changing that often (port numbers, default log level, etc), or we should make everything required.**

The former lets users specify only the things they want to change in the config initially, the latter forces everyone to be very specific about their config (basically, they'll need to start with a sample and modify it). I'm not really sure which is better here. Effectively, we're choosing between "let people easily create a ContourConfig object", vs "force people to have *all* the settings specified, and start from an example we give with the defaults set".

It could be that being more explicit is better - in that case, I think we should still set default values, since the apiserver will populate them automatically if by some chance they are empty (like when we change the spec).

In any case, regardless of what we choose, we should:
- swap all the enum types for aliased-strings with provided constants
- document allowed and default values in the docstring (so they end up in the API docs)
- Make any structs that are optional have `+optional`, `omitempty`, and be pointers.

**Comment by tsaarni** [comment]
> or we should make everything required

How would this then work when new parameters are added later? Hasn't the convention been that for backwards compatibility new parameters are always optional, allowing upgrade without "migrating" CRs at the same time?



Or would it

**Comment by tsaarni** [comment]
Also, I don't know how common this way of using Contour is, but: we run several Contour deployments per cluster and they are not upgraded at the same time.

**Comment by youngnick** [comment]
@tsaarni, these are very, very good points.



> How would this then work when new parameters are added later? Hasn't the convention been that for backwards compatibility new parameters are always optional, allowing upgrade without "migrating" CRs at the same time?

Yes, that has been the convention, and you're right, it should stay.
 
> Or would it mean that the actual up-to-date CR instance would be part of Contour example manifest and user should always apply it and keep own edits somewhere e.g. as a patch?

Yeah, I don't think this is viable.



> Also, I don't know how common this way of using Contour is, but: we run several Contour deployments per cluster and they are not upgraded at the same time.

This is your best point yet! 😄 This is a corollary of there being only able to be one apiVersion of a CRD installed at once in a cluster. That is, you can only have one `projectcontour.io/v1alpha1` `ContourConfig` at once. So, given that we need to be able to make additive, API-safe changes to the object, even after it goes GA, most things *have* to be optional with defaults. That means that we make use of the apiserver behavior I documented above to add new defaults.

That means that when we add a new field if it makes sense for it to have defaults, older configs will have the defaults added by the APIserver (since the backend version will have empty or nil values for the new field), until someone reads the field and pushes it back. There's a little bit of magic here, but it means that we can perform safe API changes and not have to have people manually add the new defaults to their spec.

It's worth taking a read of the [API Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.mds) and the [API Changes](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md) docs.

Basically, I think that we need to treat this CRD like it's a core Kubernetes API, and follow all the upstream API conventions and change rules. I think that, probably, not everything needs defaults set, and we should use the zero value for some things (TLS Ciphersuites being one) to mean "Contour should manage this", rather than "use these exact values".

I think with this I have a good understanding of some changes I think we should make, I will make a PR this afternoon so that everyone can take a look.

I should also emphasize that getting this right is *hard*, and I don't think we will get everything right this time either. In terms of release timing, we are due to cut 1.20 late next week (I'm about to update the milestone), and we *may* be able to get it in if my PR this afternoon is okay. If not, we will go with the current setup and push these changes to 1.21.

**Comment by stevesloka** [comment]
So how do you handle conflicts when a user specifies some of the config but not all that might be required? The user needs to constantly apply the change, then look at status? Does Contour default any values like today? Should we even do this or stick with the configmap. 🙃

**Comment by youngnick** [comment]
I figured it would be easier to understand this with a PR, but in short:
- kubebuilder defaults will be the same, where they can, as Contour defaults. In this PR, I've got the kubebuilder defaults defaulting to the example deployment, but we could easily change them.
- Pushing the default definitions down to lower-layer structs eliminates the problem of having a huge inline JSON in the top-level CRD definitions.

I think this exercise is very worth while, as time spent now getting this CRD's design to fit in with the conventions will save us effort later. We have very similar problems to everything I've discussed above with the configmap too. I think that having a config CRD is the way to go, but that unless we use what the community has learned about structing Kube objects, we're going to end up wasting engineering time down the road rediscovering all of the problems and solutions.

To be clear, I'm not saying the initial PR was a mistake, or that the initial work was not worthwhile. After coming back and looking again, I just think we would be foolish to not take the best lessons we can from everyone else's stuff.

As you can see, I've drafted a PR above, to see how the current state I have does against the full test suite. I'll repeat the steps I've documented above with this version once those are done, and report back here.

**Comment by youngnick** [comment]
> So how do you handle conflicts when a user specifies some of the config but not all that might be required?

I missed this.

The design principles I'm thinking of here are:
- Users should be able to specify only the bits of config they care about.
- Users should see the config they get, *including defaults* reflected in the object once they do. In some cases, like TLS Ciphersuites, because of the apiserver behaviors around defaults I discussed above, we should however *not* display the defaults. When I reviewed for the PR above, that was the only case I could see that was critical there.
- If the config the user specifies is incomplete, they should get a status message (this is a ways away yet), and the Contour install will not pick it up. (This is also a ways away).
- In terms of making changes to default values, we will need to treat them like changing field names, as they are as much a part of the API as the field names in this approach.

I am one hundred percent interested in discussion on these principles. I can see good arguments for *not* indicating defaults, and having Contour manage all the defaults (basically, that means that leaving Contour to manage things means we can update defaults at any time), but I'm currently slightly inclined to what I have above.

**Comment by skriss** [comment]
> Then, I updated the definitions for the TLS struct:
> 
> ```go
> 	// CA filename.
> 	// +optional
> 	// +kubebuilder:default="/certs/ca.crt1"
> 	CAFile string `json:"caFile,omitempty"`
> 
> 	// Client certificate filename.
> 	// +optional
> 	// +kubebuilder:default="/certs/tls.crt1"
> 	CertFile string `json:"certFile,omitempty"`
> 
> 	// Client key filename.
> 	// +optional
> 	// +kubebuilder:default="/certs/tls.key1"
> 	KeyFile string `json:"keyFile,omitempty"`
> ```
> 
> (just added a `1` to the end)
> 
> And I got:
> 
> ```yaml
> apiVersion: projectcontour.io/v1alpha1
> kind: ContourConfiguration
> metadata:
>   annotations:
>     kubectl.kubernetes.io/last-applied-configuration: |
>       {"apiVersion":"projectcontour.io/v1alpha1","kind":"ContourConfiguration","metadata":{"annotations":{},"name":"blank","namespace":"projectcontour"},"spec":{}}
>   creationTimestamp: "2022-01-18T05:59:42Z"
>   generation: 1
>   name: blank
>   namespace: projectcontour
>   resourceVersion: "456595963"
>   uid: f4940ae7-4206-4134-95e8-3467e821e279
> spec:
>   # snip
>   xdsServer:
>     address: 0.0.0.0
>     port: 9001
>     tls:
>       caFile: /certs/ca.crt1
>       certFile: /certs/tls.crt1
>       insecure: false
>       keyFile: /certs/tls.key1
>     type: contour
> ```
> 
> What this implies to me is that the actual object that's being stored is the blank one, and the defaults are being filled in for display purposes when you output it. I checked with `main` Contour, and it looks like this applies to _any_ read from the apiserver, if the original or storage object has an empty value for the object, you get the defaults, until you push that object back and change the stored version.
> 
> This means that we would get updates, _as long as they don't pull-and-reapply_ the object.
> 
> I verified this as well. Once the object has a value in the actual storage, that's what you have.
> 
> It's neat but probably not worth relying on - we would still need to tell people to update the values.

I played around with this and verified that this changing of defaults only propagates to resources if the field is a *new* field that's recently been added to the CRD and the resource in question hasn't been updated since that field was added. So for example:

1. Create a ContourConfig `config-1`, assume it does *not* have a field `foo`
2. Update the ContourConfig CRD def to include a field `foo` with default value `bar`
3. Read `config-1` from API, see it *does* show a field `foo` with value `bar`
4. Update the ContourConfig CRD def to change `foo`'s default value to `boo`
5. Read `config-1` from API again, see field `foo` now shows a value of `boo`
6. Make some update to the `config-1` CRD
7. Update the ContourConfig CRD def to change `foo`'s default value to `foobar` 
8. Read `config-1` from API again, see field `foo` still has value of `boo`; the updated default no longer propagates through

So basically, it seems that any defaults that existed on the CRD def at the time the resource was last written *do* stick; any new defaults since the last time the resource was written will be applied & can be updated until the resource is written again.

**Comment by skriss** [comment]
> In terms of making changes to default values, we will need to treat them like changing field names, as they are as much a part of the API as the field names in this approach.

I think this is probably the principle we need to keep in mind - any defaults that we define need to be considered part of the API, and changing their values will require a major version bump. So, this implies that things like TLS cipher suites, TLS versions, HTTP versions, etc. that we *expect* to change over time should *not* be defaulted in the CRD. Or, alternately, maybe they should be defaulted, and we should plan to do more frequent revs of the  API version? That certainly introduces overhead for us, but could be more explicit/clear for users. We could even put the Config CRD in a different API group than HTTPProxy, so it could be updated on a different cadence.

**Comment by skriss** [comment]
> The level of involvement required from users with regards to managing _defaults_ sounds bit unusual to me. User would need to migrate to new CR version to get new defaults. One could see defaults as something that user has handed over for implementation to decide: "I don't know or care - give me what works best". In that case, they could naturally change over time without user's needing to migrate API versions or "reset" old defaults to new defaults.

Yeah, I think ultimately I probably lean in this same direction, was just trying to brainstorm any other approaches we could use. This means we don't use kubebuilder defaults for such fields, and just rely on documentation to expose the internal defaults to users.

> Would new Contour release then still support old CRD versions as well? Or in other words, wouldn't old defaults be subject for deprecation period?

Yeah, the idea would be that Contour would support multiple API versions simultaneously, and depending on which version was used, different defaults would apply.

**Comment by youngnick** [comment]
I think that agreeing that any kubebuilder defaults are part of the API gives us two paths:
- for values we add, that we don't expect to change, _that are important for users to be able to actually see_, we add kubebuilder defaults.
- For values we add, that we either expect to change frequently, or should only need to be set in very specific cases, we have Contour silently manage defaults (still documented in the ContourConfig docs, just not as kubebuilder defaults).

I think that listen addresses and ports are good examples for kubebuilder defaults, and settings like TLS ciphersuites are better for Contour-managed defaults.

I don't think that keeping everything as kubebuilder defaults and changing API versions frequently is a good idea, there will be a *lot* of boilerplate CRD-management overhead for us in that process, for minimal gain. And it would be very unusual to have an API evolve in that way.

**Comment by youngnick** [comment]
In any case, I think that we should probably go ahead with the 1.20 release with what we have now, this is an alpha API, we are allowed to make breaking changes. We can always go to `v1alpha2` if need be.

I'll raise this in Slack so we can get agreement.

**Comment by youngnick** [comment]
You're right @tsaarni, this is why the upstream API conventions say that changing validation is always a breaking change, and changing defaults is usually a breaking change.

What do you all think about my proposal to have two classes of defaults, and use the kubebuilder defaults only for things that we're pretty confident we won't change?

**Comment by tsaarni** [comment]
Thanks for your response @youngnick,

> What do you all think about my proposal to have two classes of defaults, and use the kubebuilder defaults only for things that we're pretty confident we won't change?

Sometimes my assumptions about APIs have been proven wrong the next day, so it is difficult for me to be confident :) That might be overly cautious approach, but in my field and due to the way we deploy several independent Contour's per cluster, non-backwards changes can become rather disruptive. So, I think my conservative opinion is that using defaulting would not be worth of the additional risk.

There is clear intention that Kubernetes shall persists the defaults like they were explicitly set by user - I could learn to live with hardcoding defaults to persisted config. But it fails to deliver that promise when new fields are added. Adding an optional field was previously backwards compatible change, but because of defaulting it might not be anymore. The least that could be done, is to promise that any potential new default is a valid value for all old Contour versions. In other words, default should never select functionality that might not exist in all deployed versions of Contour. This is likely the opposite why the new default was introduced in the first place. In that case, we'd need to step the CRD version, but I dont know if we will be alert enough to realize when this must be done.

**Comment by youngnick** [comment]
The thing I'm trying to avoid with some defaulting is that you have a ContourConfig that looks like this:
```yaml
  spec: {}
```
which would probably be the case for the example deployment. That doesn't seem like a great user experience for the ContourConfig, although I suppose the idea being "the only config that you list is what you change" is not that hard a sell.

**Comment by youngnick** [comment]
Ping here, it seems like our feeling is that we would be better off having the ContourConfig work like this:
- anything you list is what you want to override Contour's defaults with
- Defaults are listed in the docs on projectcontour.io

There's no way to do something like what we have with the config file, where you list the default values commented out in the example file, sadly.

My only concern here is that it makes the defaults not very discoverable, but maybe this is not a big deal?

@tsaarni, what do you think?

**Comment by skriss** [comment]
Yep, I'm in agreement with the above as a path forward given all the challenges we've uncovered here.

As discussed, my one suggestion is to capture the internal default values in the godoc for each ContourConfiguration field, since this will be visible not only in projectcontour.io, but also in the output of `kubectl explain`, which should be helpful.

**Comment by tsaarni** [comment]
Yes, I agree as well :+1:

**Comment by youngnick** [comment]
Okay, it seems like consensus here is this:
- We remove all the kubebuilder defaults from the ContourConfig CRD. Not validation, just default values. Everything should still be optional though.
- We replace them with Contour-managed defaults that are documented in the relevant godoc, so that it shows up on the website and in `kubectl describe`.
- We provide an example YAML for ContourConfig with everything commented out.

Is that everything? Can I get some +1s or other reactions from everyone here? Once we all agree, I think we can move this forward and get the ContourConfig bits sorted out and brought to beta or even GA.

**Comment by sunjayBhatia** [comment]
+1

**Comment by youngnick** [comment]
Okay, I'll go back to my PR and see what I can do.

**Comment by sunjayBhatia** [comment]
@skriss I think this is all done? based on your refactors etc.

**Comment by skriss** [comment]
> @skriss I think this is all done? based on your refactors etc.

The one thing I haven't gotten to is removing all of the kubebuilder enum validations and moving that logic into the controllers. I think I have a WIP branch with it but I'll have to dust it off, probably isn't gonna make 1.21 at this point.

Otherwise, we could probably close this out and do another review if/when we want to move these to GA.

**Comment by youngnick** [comment]
I agree when can close this out for 1.21.

**Comment by skriss** [comment]
Filed https://github.com/projectcontour/contour/issues/4510.
